text/0000-supertrait-item-shadowing-v2.md

```diff
@@ -0,0 +1,152 @@
+- Feature Name: `supertrait_item_shadowing`
+- Start Date: 2024-05-04
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+When name resolution encounters an ambiguity between 2 trait methods when both traits are in scope, if one trait is a sub-trait of the other then select that method instead of reporting an ambiguity error.
+
+# Motivation
+[motivation]: #motivation
+
+
+The libs-api team would like to stabilize `Iterator::intersperse` but has a problem. The `itertools` crate already has:
+
+```rust
+// itertools
+trait Itertools: Iterator {
+    fn intersperse(self, element: Self::Item) -> Intersperse<Self>;
+}
+```
+
+This method is used in crates with code similar to the following:
+
+```rust
+use core::iter::Iterator; // Implicit import from prelude
+
+use itertools::Itertools as _;
+
+fn foo() -> impl Iterator<Item = &'static str> {
+    "1,2,3".split(",").intersperse("|")
+    // ^ This is ambiguous: it could refer to Iterator::intersperse or Itertools::intersperse
+}
+```
+
+This code actually works today because `intersperse` is an unstable API, which works because the compiler already has [logic](https://github.com/rust-lang/rust/pull/48552) to prefer stable methods over unstable methods when an amiguity occurs.
```

Comment:
```suggestion
This code actually works today because `intersperse` is an unstable API, which works because the compiler already has [logic](https://github.com/rust-lang/rust/pull/48552) to prefer stable methods over unstable methods when an ambiguity occurs.
```